python/GafferUI/PathListingWidget.py

```diff
@@ -776,6 +782,74 @@ def __dragBegin( self, widget, event ) :
 	def __dragEnd( self, widget, event ) :
 
 		GafferUI.Pointer.setCurrent( None )
+		self.__dragBeginIndex = None
+		self.__currentDragIndex = None
+
+	def __dragEnter( self, widget, event ) :
+
+		self.__currentDragIndex = self.__indexAt( event.line.p0 )
+		if self.__currentDragIndex is None :
+			return False
+
+		if self.getColumns()[self.__currentDragIndex.column()].dragEnterSignal()(
+			self.__pathForIndex( self.__currentDragIndex ), self, event
+		) :
+			self.__setHighlightedIndex( self.__currentDragIndex )
+			return True
+
+	def __dragMove( self, widget, event ) :
+
+		previousIndex = self.__currentDragIndex
+		self.__currentDragIndex = self.__indexAt( event.line.p0 )
+
+		# We've moved out of the previous index
+		if previousIndex is not None and previousIndex != self.__currentDragIndex :
+			self.__setHighlightedIndex( None )
+			self.getColumns()[previousIndex.column()].dragLeaveSignal()(
+				self.__pathForIndex( previousIndex ), self, event
+			)
+
+		if self.__currentDragIndex is None or self.__currentDragIndex == self.__dragBeginIndex :
+			self.__setHighlightedIndex( None )
+			return False
+
+		# We've moved into a new index
+		if previousIndex is None or previousIndex != self.__currentDragIndex :
+			if self.getColumns()[self.__currentDragIndex.column()].dragEnterSignal()(
+				self.__pathForIndex( self.__currentDragIndex ), self, event
+			) :
+				self.__setHighlightedIndex( self.__currentDragIndex )
+				return True
+		elif previousIndex == self.__currentDragIndex :
+			# We've moved within the same index
+			return self.getColumns()[self.__currentDragIndex.column()].dragMoveSignal()(
+				self.__pathForIndex( self.__currentDragIndex ), self, event
+			)
+
+		return False
+
+	def __dragLeave( self, widget, event ) :
+
+		self.__setHighlightedIndex( None )
+
+		if self.__currentDragIndex is not None :
+			return self.getColumns()[self.__currentDragIndex.column()].dragLeaveSignal()(
+				self.__pathForIndex( self.__currentDragIndex ), self, event
+			)
+
+		return False
+
+	def __drop( self, widget, event ) :
+
+		self.__setHighlightedIndex( None )
+
+		index = self.__indexAt( event.line.p0 )
+		if index is None or index == self.__dragBeginIndex :
+			return False
+
+		return self.getColumns()[index.column()].dropSignal()(
```

Comment:
As far as I can see, we end up calling `drop` on the column under the mouse here, even if that column would have rejected `dragEnter`. I think maybe we need `__currentDragIndex` to track the last column to accept the drag (and be None when the current column rejects the drag) and then deliver drop to that index here?

Reply:
Likewise I think we're calling `dragLeaveSignal()` on columns that didn't accept `dragEnterSignal()`.

Reply:
Reworked this so `__currentDragIndex` only tracks the last column to accept the drag in a62e4ee8d2f8152426e9a44c8c5fcfc25f516c75.